### PR TITLE
Check XBE signature on Open

### PR DESCRIPTION
**Pull request by GXTX**
Check XBE signature on open & allow users to disable this pop-up using the setting `allowAdminPrivilege`.

I think this will produce better overall compatibility reports of games as it incentivizes users to use unmodified games.

### CODE REVIEW

File: src/gui/WndMain.cpp [hunk 1]
```diff
@@ -1990,7 +1990,16 @@ void WndMain::OpenXbe(const char *x_filename)
 
         return;
     }
-
+	
+	if (!m_Xbe->CheckXbeSignature() && !g_Settings->m_core.allowAdminPrivilege)
```
Comment (JayFoxRox) [inline src/gui/WndMain.cpp]: Checks should be done in different order: Only run check if `!g_Settings->m_core.allowAdminPrivilege`.

Currently, you always check, but only decide if you want a messagebox if there's no admin privileges (possibly discarding the check result).

File: src/gui/WndMain.cpp [hunk 2]
```diff
@@ -1990,7 +1990,16 @@ void WndMain::OpenXbe(const char *x_filename)
...
     }
-
+	
+	if (!m_Xbe->CheckXbeSignature() && !g_Settings->m_core.allowAdminPrivilege)
+	{
+		int ret = MessageBox(m_hwnd, "XBE signature check failed!\n"
+			"\nThis is dangerous, as maliciously modified Xbox titles could take control of your system.\n"
+			"\nAre you sure you want to continue?", "Cxbx-Reloaded", MB_ICONEXCLAMATION | MB_YESNO);
+		if (ret != IDYES)
+			return;
```
Comment (JayFoxRox) [inline src/gui/WndMain.cpp]: What happens to allocations like `m_Xbe`?

See error handling in the `if(m_Xbe->HasError())` case, for what's necessary to abort.

**Comment by JayFoxRox** [comment]
While a good idea on paper, this feels like a bad idea in practice.

At least `CheckXbeSignature` should check all valid / common signatures: MS retail, MS debug, MS chihiro.
I believe some Xbox Live betas also had custom signatures.

Currently, the warning is a no-op and gives a false sense of security. Because it allows zero-signatures from homebrew, which *can* be malicious, because it wasn't verified by a trusted source.
I believe there's hardly any XBE which does not succeed the check in the current form, and if it does, it's probably not malicious, but just corrupted.
Meanwhile, even malicious XBEs are allowed.

So the homebrew signatures, like habibi or a zero signature, would have to be allowed, but warned.

I'm also not sure about the signature specifics, but if it's also checking header fields like "allowed-media-flags", then almost any XBE in existence will now be reported (affecting usability, and scaring people away). Many dumping solutions will modify these flags. To my knowledge, some tools also patched code to work around security checks.

File: src/gui/WndMain.cpp (hunk 1 above)
Comment (GXTX) [reply src/gui/WndMain.cpp]: Fixed

File: src/gui/WndMain.cpp (hunk 2 above)
Comment (GXTX) [reply src/gui/WndMain.cpp]: Fixed

**Comment by GXTX** [comment]
Went ahead and fixed your concerns. As to the last paragraph there, like we chatted about sometime last night the biggest reason for this PR is users are currently for the most part using non-original versions of games and I believe this pollutes the compatibility tracker immensely by not only having different versions of games but also different flags for each game. There also comes a risk in using these modified versions such as if a bad actor were to inject malicious code after a user uses closed source tools to create personal copies.

If the user doesn't know how to set these tools not to modify game code I think that's a separate issue we need to address, and I think runs back into your concerns with the redump project -- they aren't saving the games correctly with ALL relevant information.

File: src/gui/WndMain.cpp [hunk 3]
```diff
@@ -1991,13 +1991,21 @@ void WndMain::OpenXbe(const char *x_filename)
...
+	if (!g_Settings->m_core.allowAdminPrivilege && !m_Xbe->CheckXbeSignature())
 	{
 		int ret = MessageBox(m_hwnd, "XBE signature check failed!\n"
 			"\nThis is dangerous, as maliciously modified Xbox titles could take control of your system.\n"
 			"\nAre you sure you want to continue?", "Cxbx-Reloaded", MB_ICONEXCLAMATION | MB_YESNO);
 		if (ret != IDYES)
+		{
+			delete m_Xbe; m_Xbe = nullptr;
+			
+			RedrawWindow(m_hwnd, NULL, NULL, RDW_INVALIDATE);
```
Comment (JayFoxRox) [inline src/gui/WndMain.cpp]: nit: order in other case frees and redraws first, then shows message-box.

**Comment by RadWolfie** [comment]
`CheckXbeSignature` is using XePublicKeyData which is already provided in the source code. I'm not sure if debug kit (most likely yes) and chihiro are using the same public key. If not, then someone will have to get the RSA public key from the hardware.

> I believe some Xbox Live betas also had custom signatures.

I'm not sure about this since I don't have access to any of Xbox Live games. Plus it should produce a failure on real hardware if it was using a different RSA public key.

**Comment by RadWolfie** [comment]
Other than that, looks good to me since it will popup only on GUI's end when open an xbe file.

**Comment by LukeUsher** [comment]
Thanks!

Currently this shows for Chihiro executables, but that's because we don't have the Chihiro xbe key in Cxbx-R yet. I'll address that in a later commit.

**Pull request merged**